File: packages/python/fractary_codex/references/resolver.py
"""
Reference resolution utilities.

This module provides functions to resolve codex:// URIs to actual file paths
and detect the current project context from git repositories.
"""

import os
import re
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from ..errors import ReferenceError, ValidationError
from .parser import ParsedReference, parse_reference


@dataclass
class ResolvedReference:
    """Result of resolving a codex:// reference.

    Attributes:
        reference: The original parsed reference
        local_path: Local filesystem path (if resolved locally)
        remote_url: Remote URL (if resolved to remote storage)
        cache_path: Path in the local cache (if cached)
        source: How the reference was resolved ('local', 'cache', 'remote')
    """

    reference: ParsedReference
    local_path: Optional[Path] = None
    remote_url: Optional[str] = None
    cache_path: Optional[Path] = None
    source: str = "unknown"


@dataclass
class ProjectContext:
    """Current project context detected from environment.

    Attributes:
        org: Organization name
        project: Project name
        root_path: Path to project root
        git_remote: Git remote URL (if detected)
    """

    org: str
    project: str
    root_path: Path
    git_remote: Optional[str] = None


def detect_current_project(
    start_path: Optional[Path] = None,
) -> Optional[ProjectContext]:
    """Detect the current project context from git repository.

    This function walks up the directory tree looking for a .git directory,
    then extracts organization and project information from the git remote.

    Args:
        start_path: Starting directory (defaults to current working directory)

    Returns:
        ProjectContext if detected, None otherwise

    Examples:
        >>> ctx = detect_current_project()
        >>> if ctx:
        ...     print(f"Project: {ctx.org}/{ctx.project}")
    """
    if start_path is None:
        start_path = Path.cwd()

    start_path = Path(start_path).resolve()

    # Find git root
    git_root = _find_git_root(start_path)
    if git_root is None:
        return None

    # Get remote URL
    remote_url = _get_git_remote(git_root)
    if remote_url is None:
        # Fall back to directory name
        return ProjectContext(
            org="local",
            project=git_root.name,
            root_path=git_root,
            git_remote=None,
        )

    # Parse org/project from remote URL
    org, project = _parse_remote_url(remote_url)

    return ProjectContext(
        org=org,
        project=project,
        root_path=git_root,
        git_remote=remote_url,
    )


def resolve_reference(
    ref: ParsedReference | str,
    *,
    base_dir: Optional[Path] = None,
    cache_dir: Optional[Path] = None,
    project_context: Optional[ProjectContext] = None,
) -> ResolvedReference:
    """Resolve a codex:// reference to a file path or URL.

    This function attempts to resolve the reference in the following order:
    1. Local filesystem (if org/project matches current project)
    2. Local cache (if available)
    3. Remote URL (GitHub raw content)

    Args:
        ref: Parsed reference or URI string
        base_dir: Base directory for local resolution
        cache_dir: Cache directory for cached content
        project_context: Current project context (auto-detected if not provided)

    Returns:
        ResolvedReference with resolution details

    Raises:
        ValidationError: If URI is invalid
        ReferenceError: If reference cannot be resolved
    """
    # Parse if string
    if isinstance(ref, str):
        ref = parse_reference(ref)

    # Auto-detect project context
    if project_context is None:
        project_context = detect_current_project(base_dir)

    # Try local resolution first
    if project_context and _matches_current_project(ref, project_context):
        local_path = project_context.root_path / ref.path
        if local_path.exists():
            return ResolvedReference(
                reference=ref,
                local_path=local_path,
                source="local",
            )

    # Try cache
    if cache_dir:
        cache_path = _get_cache_path(ref, cache_dir)
        if cache_path.exists():
            return ResolvedReference(
                reference=ref,
                cache_path=cache_path,
                source="cache",
            )

    # Build remote URL
    remote_url = _build_github_url(ref)

    return ResolvedReference(
        reference=ref,
        remote_url=remote_url,
        source="remote",
    )


def resolve_references(
    refs: list[ParsedReference | str],
    *,
    base_dir: Optional[Path] = None,
    cache_dir: Optional[Path] = None,
) -> list[ResolvedReference]:
    """Resolve multiple references at once.

    Args:
        refs: List of references to resolve
        base_dir: Base directory for local resolution
        cache_dir: Cache directory

    Returns:
        List of resolved references
    """
    project_context = detect_current_project(base_dir)
    return [
        resolve_reference(
            ref,
            base_dir=base_dir,
            cache_dir=cache_dir,
            project_context=project_context,
        )
        for ref in refs
    ]


def _find_git_root(start_path: Path) -> Optional[Path]:
    """Find the git repository root by walking up the directory tree."""
    current = start_path

    while current != current.parent:
        git_dir = current / ".git"
        if git_dir.exists():
            return current
        current = current.parent

    # Check root directory
    if (current / ".git").exists():
        return current

    return None


def _get_git_remote(git_root: Path, remote_name: str = "origin") -> Optional[str]:
    """Get the URL of a git remote."""
    try:
        result = subprocess.run(
            ["git", "remote", "get-url", remote_name],
            cwd=git_root,
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            return result.stdout.strip()
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        pass
    return None


def _parse_remote_url(url: str) -> tuple[str, str]:
    """Parse organization and project from a git remote URL.

    Supports:
    - SSH: git@github.com:org/project.git
    - HTTPS: https://github.com/org/project.git
    - GitLab, Bitbucket, etc.
    """
    # SSH format: git@github.com:org/project.git
    ssh_match = re.match(r"git@[^:]+:([^/]+)/([^/]+?)(?:\.git)?$", url)
    if ssh_match:
        return ssh_match.group(1), ssh_match.group(2)

    # HTTPS format: https://github.com/org/project.git
    https_match = re.match(r"https?://[^/]+/([^/]+)/([^/]+?)(?:\.git)?$", url)
    if https_match:
        return https_match.group(1), https_match.group(2)

    # Fallback: use URL parts
    parts = url.rstrip("/").split("/")
    if len(parts) >= 2:
        project = parts[-1].replace(".git", "")
        org = parts[-2]
        return org, project

    return "unknown", "unknown"


def _matches_current_project(ref: ParsedReference, ctx: ProjectContext) -> bool:
    """Check if a reference matches the current project context."""
    return ref.org.lower() == ctx.org.lower() and ref.project.lower() == ctx.project.lower()


def _get_cache_path(ref: ParsedReference, cache_dir: Path) -> Path:
    """Get the cache file path for a reference."""
    return cache_dir / ref.org / ref.project / ref.path


def _build_github_url(ref: ParsedReference, branch: str = "main") -> str:
    """Build a GitHub raw content URL for a reference."""
    return f"https://raw.githubusercontent.com/{ref.org}/{ref.project}/{branch}/{ref.path}"

File: sdk/js/src/references/resolver.ts
/**
 * URI resolution for codex references
 *
 * Resolves codex:// URIs to filesystem paths, determining whether
 * the reference points to the current project (local file) or
 * needs to be fetched from cache/remote.
 */

import path from 'path'
import { execSync } from 'child_process'
import { parseReference, type ParsedReference } from './parser.js'
import type { StorageProviderType } from '../storage/provider.js'

/**
 * Resolved reference with filesystem paths
 */
export interface ResolvedReference extends ParsedReference {
  /** Path in the cache directory */
  cachePath: string
  /** Whether this URI refers to the current project */
  isCurrentProject: boolean
  /** Local filesystem path if current project */
  localPath?: string
  /** Source type for storage routing */
  sourceType?: StorageProviderType
  /** File plugin source name if sourceType is 'file-plugin' */
  filePluginSource?: string
}

/**
 * Options for resolving references
 */
export interface ResolveOptions {
  /** Base directory for cache (default: .fractary/codex/cache) */
  cacheDir?: string
  /** Override current organization detection */
  currentOrg?: string
  /** Override current project detection */
  currentProject?: string
  /** Current working directory (for git detection) */
  cwd?: string
  /** Unified config for file plugin source resolution */
  config?: any // Type will be UnifiedConfig, using 'any' to avoid circular dependency
}

/**
 * Default cache directory (v4.0 standard)
 */
export const DEFAULT_CACHE_DIR = '.fractary/codex/cache'

/**
 * Detect the current project from git remote
 *
 * @param cwd - Current working directory
 * @returns Object with org and project, or nulls if not detected
 */
export function detectCurrentProject(cwd?: string): { org: string | null; project: string | null } {
  try {
    const options = cwd ? { cwd, encoding: 'utf-8' as const } : { encoding: 'utf-8' as const }
    const stdout = execSync('git remote get-url origin', options)
    const url = stdout.toString().trim()

    // Parse GitHub URLs (SSH and HTTPS formats)
    // git@github.com:org/repo.git
    // https://github.com/org/repo.git
    // https://github.com/org/repo
    const patterns = [
      /github\.com[:/]([^/]+)\/([^/]+?)(?:\.git)?$/,
      /gitlab\.com[:/]([^/]+)\/([^/]+?)(?:\.git)?$/,
      /bitbucket\.org[:/]([^/]+)\/([^/]+?)(?:\.git)?$/,
    ]

    for (const pattern of patterns) {
      const match = url.match(pattern)
      if (match && match[1] && match[2]) {
        return { org: match[1], project: match[2] }
      }
    }
  } catch {
    // Git not available or not in a git repo
  }

  return { org: null, project: null }
}

/**
 * Get the current organization and project
 *
 * Priority:
 * 1. Explicit options
 * 2. Environment variables (CODEX_CURRENT_ORG, CODEX_CURRENT_PROJECT)
 * 3. Git remote detection
 *
 * @param options - Resolve options
 * @returns Object with org and project
 */
export function getCurrentContext(options: ResolveOptions = {}): {
  org: string | null
  project: string | null
} {
  // Priority 1: Explicit options
  if (options.currentOrg && options.currentProject) {
    return { org: options.currentOrg, project: options.currentProject }
  }

  // Priority 2: Environment variables
  const envOrg = process.env.CODEX_CURRENT_ORG
  const envProject = process.env.CODEX_CURRENT_PROJECT
  if (envOrg && envProject) {
    return { org: envOrg, project: envProject }
  }

  // Priority 3: Git detection
  return detectCurrentProject(options.cwd)
}

/**
 * Calculate the cache path for a URI
 *
 * @param org - Organization
 * @param project - Project
 * @param filePath - File path within project
 * @param cacheDir - Base cache directory
 * @returns The cache file path
 */
export function calculateCachePath(
  org: string,
  project: string,
  filePath: string,
  cacheDir: string
): string {
  // Cache structure: {cacheDir}/{org}/{project}/{path}
  return path.join(cacheDir, org, project, filePath)
}

/**
 * Resolve a codex URI to filesystem paths
 *
 * @param uri - The URI to resolve
 * @param options - Resolution options
 * @returns The resolved reference or null if invalid
 */
export function resolveReference(
  uri: string,
  options: ResolveOptions = {}
): ResolvedReference | null {
  const parsed = parseReference(uri)
  if (!parsed) {
    return null
  }

  const cacheDir = options.cacheDir || DEFAULT_CACHE_DIR
  const currentContext = getCurrentContext(options)

  // Calculate cache path
  const cachePath = parsed.path
    ? calculateCachePath(parsed.org, parsed.project, parsed.path, cacheDir)
    : path.join(cacheDir, parsed.org, parsed.project)

  // Check if this is the current project
  const isCurrentProject =
    currentContext.org === parsed.org && currentContext.project === parsed.project

  // Build the resolved reference
  const resolved: ResolvedReference = {
    ...parsed,
    cachePath,
    isCurrentProject,
  }

  // If it's the current project and there's a path, provide the local path
  if (isCurrentProject && parsed.path) {
    resolved.localPath = parsed.path

    // Check if this path belongs to a file plugin source
    if (options.config) {
      const fileSource = detectFilePluginSource(parsed.path, options.config)
      if (fileSource) {
        resolved.sourceType = 'file-plugin'
        resolved.filePluginSource = fileSource.name
        resolved.localPath = fileSource.fullPath
      }
    }
  }

  return resolved
}

/**
 * Detect if a path belongs to a file plugin source
 *
 * @param filePath - The file path from the URI
 * @param config - Unified configuration
 * @returns Matching file source info or null
 */
function detectFilePluginSource(
  filePath: string,
  config: any
): { name: string; fullPath: string } | null {
  if (!config?.file?.sources) {
    return null
  }

  // Check each file source to see if the path matches
  for (const [sourceName, source] of Object.entries(config.file.sources) as [string, any][]) {
    const basePath = source.local?.base_path
    if (!basePath) {
      continue
    }

    // Normalize paths for comparison
    const normalizedPath = filePath.replace(/^\.\//, '').replace(/^\//, '')
    const normalizedBasePath = basePath.replace(/^\.\//, '').replace(/^\//, '').replace(/\/$/, '')

    // Check if path starts with source's base_path
    // Examples:
    //   filePath: "specs/SPEC-001.md"
    //   basePath: ".fractary/specs"
    //   normalizedPath: "specs/SPEC-001.md"
    //   normalizedBasePath: ".fractary/specs"
    //
    // Or if path is already prefixed:
    //   filePath: ".fractary/specs/SPEC-001.md"
    //   normalizedPath: ".fractary/specs/SPEC-001.md"
    //
    // We need to check two scenarios:
    // 1. Path already includes base_path: ".fractary/specs/SPEC-001.md" starts with ".fractary/specs"
    // 2. Path is relative to base_path: "specs/SPEC-001.md" and base_path is ".fractary/specs"
    //    In this case, we extract the source name from base_path and check if path starts with it

    if (normalizedPath.startsWith(normalizedBasePath)) {
      // Scenario 1: Path already includes full base_path
      return {
        name: sourceName,
        fullPath: path.join(basePath, normalizedPath.substring(normalizedBasePath.length).replace(/^\//, '')),
      }
    }

    // Scenario 2: Extract source name from base_path (e.g., "specs" from ".fractary/specs")
    const sourceNameInPath = normalizedBasePath.split('/').pop()
    if (sourceNameInPath && normalizedPath.startsWith(sourceNameInPath + '/')) {
      // Path starts with source name (e.g., "specs/SPEC-001.md")
      return {
        name: sourceName,
        fullPath: path.join(basePath, normalizedPath),
      }
    }
  }

  return null
}

/**
 * Resolve multiple URIs at once
 *
 * @param uris - Array of URIs to resolve
 * @param options - Resolution options
 * @returns Array of resolved references (nulls filtered out)
 */
export function resolveReferences(
  uris: string[],
  options: ResolveOptions = {}
): ResolvedReference[] {
  return uris.map((uri) => resolveReference(uri, options)).filter((r): r is ResolvedReference => r !== null)
}

/**
 * Check if a URI refers to the current project
 *
 * @param uri - The URI to check
 * @param options - Resolution options
 * @returns true if the URI refers to the current project
 */
export function isCurrentProjectUri(uri: string, options: ResolveOptions = {}): boolean {
  const resolved = resolveReference(uri, options)
  return resolved?.isCurrentProject ?? false
}

/**
 * Get the relative cache path for a URI (without the cache directory prefix)
 *
 * @param uri - The URI
 * @returns The relative path or null if invalid
 */
export function getRelativeCachePath(uri: string): string | null {
  const parsed = parseReference(uri)
  if (!parsed) {
    return null
  }

  if (parsed.path) {
    return path.join(parsed.org, parsed.project, parsed.path)
  }
  return path.join(parsed.org, parsed.project)
}
